### ISSUE TITLE
Infinite loop when one or more failed new records present

### ISSUE THREAD

**Issue by gduprey** [godaddy/activerecord-delay_touching#14]
When using this on a block of record creates where one or more of the creates failed (and so have no id value -- they are still nil), delay_touching goes into an infinite loop.

Looking at the code, it appears that the problem is that the key for the failed record is nil and in the State classes updated() method, attempts to delete that from the @records hash/set are ignored and so the hash/set never empties out.  As a result, the loop that attempts to iteratively process all pending touches never completes.

Perhaps a check on .persisted? should be done and if false, silently skipped (since the problem is an unending attempt to set the time stamps on a record with an ID of NULL).

**Comment by mtuckergd** [comment]
Hi @gduprey, I'll try to find some time to look deeper into this. Do you happen to have a test case that shows the issue? That would expedite finding a resolution.
Thanks for your help!

**Comment by goodwill** [comment]
Same here, please see if you can introduce a fix

**Comment by krainboltgreene** [comment]
We encountered this issue, it occurs when two things happen:

1. A touch is scheduled (specifically in our case on a `belongs_to`)
2. A `rollback` is fired

What happens is that the rollback succeeds, but now it's time to delayed touch, but in this case it's touching `id = null` because *the record got rolled back*.

**Comment by krainboltgreene** [comment]
My guess is it's in this definition of `more_present?`

https://github.com/godaddy/activerecord-delay_touching/blob/master/lib/activerecord/delay_touching.rb#L63-L71

**Comment by krainboltgreene** [comment]
Due to the fact that it never has a primary key, I'm betting the equality check for `subtract` fails to actually remove the records from the set, which causes for an infinite loop.

**Comment by oehlschl** [comment]
So, I've been discussing this issue offline with @krainboltgreene.

TL;DR, in the case of rollbacks, ActiveRecord lazily syncs in-memory objects with transaction state:

https://github.com/rails/rails/blob/5d1402a1011f58b405e42007d3ceed4e122d273e/activerecord/lib/active_record/persistence.rb#L86-L86

Specifically, `id` attributes are unset in memory, which breaks the Set membership/equality checks in the DelayTouching State object until the keys are reset/rehashed. This functionality wasn't introduced into Set until Ruby 2.5, though it can be shimmed easily enough:

https://ruby-doc.org/stdlib-2.5.1/libdoc/set/rdoc/Set.html#method-i-reset

(See also https://ruby-doc.org/core-2.5.1/Hash.html#method-i-rehash)

I think the best answer is to manually introduce some `sync_with_transaction_state` logic into DelayTouching, and the only questions are where (performance / frequency considerations) and, to a lesser extent, how (to support differented versions of Ruby and Rails).

I suggested something like the following in addition to `DelayTouching::State`, to be invoked by either `state.records_by_attrs_and_class` or more globally (and frequently) by `DelayTouching.state`:

```ruby
# ActiveRecord objects need to be synced with transaction state in the case of a rollback.
# Because they are used as hash keys and the records may change, we also need to reset/rehash the State.
def sync_with_transaction_state
  @records.each do |_attrs, records|
    # TODO possibly invoke a public method that may be more future proof (.peristed?)
    # NOTE: sync_with_transaction_state exists in both Rails 4 and 5
    records.each {|record| record.send(:sync_with_transaction_state) }

    # Set#reset only exists >= Ruby 2.5
    # TODO possibly shim?
    records.instance_variable_get(:@hash).rehash
  end
end
```

Thoughts? (@krainboltgreene et al.)

**Comment by oehlschl** [comment]
Also, to replicate:

```ruby
ActiveRecord::Base.delay_touching do
  ActiveRecord::Base.transaction do

    # write and touch any new record
    record = Post.create!
    record.touch
    raise ActiveRecord::Rollback
  end

  # results in an infinite loop
end
```

**Comment by mtuckergd** [comment]
Hey folks. Thanks so much for digging into this! I'd love to work together to narrow down a good solution.
Let me get the repro plugged into a proper failing unit test (i.e. one that detects the infinite loop without _actually_ looping infinitely) and we can go from there. Let me know if you have more thoughts on the solution, in the meantime.

**Comment by mtuckergh** [comment]
This is fixed as of e71ade4. Version 1.1.0 should be available on rubygems sometime today.

HUGE thanks to @krainboltgreene and @oehlschl for digging in to identify the root cause and flesh out viable solutions!

**Comment by mtuckergh** [comment]
https://rubygems.org/gems/activerecord-delay_touching/versions/1.1.0
